Add #[documented_function] attribute macro

SeseMueller opened the pull request:
Implements #12. 

The current attribute macro simply adds a const variable of type `&'static str` named like the function, but with added `_docs`, after the function. 

Example from test file: 
```Rust
use documented::documented_function;

#[test]
fn it_works() {
    /// This is a test function
    #[documented_function]
    #[allow(dead_code)]
    fn test_fn() {}

    assert_eq!(test_fn_docs, "This is a test function");
}
```

The variable has the same visibility (pub, pub(crate), etc.) of the function. 
The content of the variable works just like the rest of the crate, as I use the same function to extract the docstring.
The macro also suppresses the lint that it's a const that doesn't use upper case. 


What's the way forward now? I don't fully understand the config part of the crate yet.

cyqsimon commented:
Thanks for the PR, I've got a few comments.

> The current attribute macro simply adds a const variable of type `&'static str` named like the function, but with added `_docs`, after the function.

I think it's best if we use SCREAMING_SNAKE_CASE for the generated constants by default, as per Rust code style convention. Think about using this constant at the call site; it would be weird to see a lowercase variable that is actually a constant. We can use config options to allow customisations, which brings us neatly to the config part.

> What's the way forward now? I don't fully understand the config part of the crate yet.

The config code probably looks very weird at first glance, but the style is actually just what the authors of `syn` [recommends](https://docs.rs/syn/latest/syn/macro.custom_keyword.html). It's an easy (well, relatively easy) way to define configurable options for use with proc macros.

If you wish to experiment and try implementing this part by yourself, please go ahead and take as much time as you need; alternatively if you no longer want to work on it, inform me at any time and I'll be happy to take it from there. Thanks.

SeseMueller commented:
I've now changed the case and implemented the config reading, as you suggested. 
For the function, it's just for the main docs as there is are no sub-docs.

The syntax of the attribute macro was a bit complicated, but the existing `Parse` implementation for `ConfigCustomisations` you already used made things very easy. 

I also added the documentation for the macro as well as doctests, which are passing. 

Could you please check the test cases?
If I missed nothing, I believe this should be able to be deployed.

cyqsimon commented:
Sorry for having left this PR hanging; I had lots of work in the past two weeks so I was a bit strapped for time.

Code looks pretty good in general. There are three main things I have in mind:

1. I still want the ability to set a custom constant name. I feel like it's important. Although I'm still a bit undecided on where or how to add this option.
2. Given that this addition makes the crate "no longer just derive macros", I think it makes sense to do some renaming. I'm thinking of retiring the `documented-derive` crate name and renaming to something like `documented-macros`. We would also have to update README, project description, and various bits and bobs.
3. When I come to think about it, there's really no reason we should limit the use of this new attribute macros to functions. I could imagine that someone would find a "detached" doc constant useful on a struct, trait, etc.

cyqsimon commented:
I'll start with 2 and 3 on this branch. We can leave 1 til last.

SeseMueller commented:
1 could be done with attributes (I think they are called, it's been some time)
So then the user could do `#[documented_function(name=docs_of_fn)]` for example. 
I'll start to work on it.

cyqsimon commented:
That's what `ConfigCustomisations` is doing. The problem is, this struct is currently serving everything, derive macros and this new attribute macro. But obviously this custom name argument is only applicable to the attribute macro, so if the user attempts to use this argument with derive macros, we should produce an error.

So I was just wondering what's the cleanest way of doing this. Should I:
- Split `ConfigCustomisations` into two different structs - one for derive macros and one for attribute macros?
- Or should I still use `ConfigCustomisations` for both purposes but add logic during parsing (or whatever) to handle this?

Honestly neither option seems very appealing to me from a cleaniness and maintainability perspective. You know, that feeling of being stuck between a rock and a hard place. So I guess I really have to explore both options to an informed opinion on which one is the lesser evil. I'm happy to hear your input on this matter too.

SeseMueller commented:
I think that we should stick to what differentiates this macro from the other, existing macros: how the output is read. 
As to the macro itself, if I see it correctly, it should not be entirely difficult to move from functions to everything else that supports documentation. 
Then writing a new variable/creating a field would be the differentiator between this macro and the existing ones. Both options aren't great, but I believe creating a new Customization would be a better idea. 

Should I wait for you to implement the Customization?

cyqsimon commented:
> Then writing a new variable/creating a field would be the differentiator between this macro and the existing ones. Both options aren't great, but I believe creating a new Customization would be a better idea.

Sounds reasonable; I'll try this approach first.

> Should I wait for you to implement the Customization?

Yes please. This should be rather quick.

cyqsimon commented:
Turns out, it's slightly more complicated than I thought, and it requires a bit more reorganisation. And again, work is busier than I expected 🙁. I'll get to it tomorrow, hopefully.

cyqsimon commented:
Okay, customisations are implemented now. I haven't added/updated the tests yet because I imagine that's most efficiently done after we're finished with 3:

> When I come to think about it, there's really no reason we should limit the use of this new attribute macros to functions. I could imagine that someone would find a "detached" doc constant useful on a struct, trait, etc.

I took a quick look at the `syn` APIs, and it doesn't seem as trivial as I thought. Can you please take a look?

SeseMueller commented:
Ok, so I didn't look at it that close either.
The syn crate has a file `item.rs` where the structure of every possible item is defined. There's quite a lot, but reading my current code, we basically just need it to have attributes, a name, a visibility and be parsable by syn. Instead of using the `syn::parse_macro_input!` macro, the macro description tells us it's just a convenience instead of using `syn::parse::<Type>` and handling the error automatically. 

For every item we want to support annotating, we need to parse it, extract the attributes, name and visibility and then do the same stuff my current code does. 

I hope I'm not overlooking anything. I'll see when I can take the time to work on it.

cyqsimon commented:
Yeah this looks like quite the pain.

https://docs.rs/syn/latest/syn/enum.Item.html#variants

I hope there's a better way to do this than matching all variants we want to support.

cyqsimon commented on documented-test/src/attr/docs_const.rs:
```diff
@@ -0,0 +1,86 @@
+    assert_eq!(TEST_FN_DOCS, "This is a test function\nThis is the second line of the doc\nThis is the third line of the doc");
+}
+
+#[test]
+fn parameters_work() {
+    /// This is a test function
+    #[docs_const]
+    #[allow(dead_code)]
+    fn test_fn(_test: String) {}
+
+    assert_eq!(TEST_FN_DOCS, "This is a test function");
+}
+
+#[test]
+fn return_and_body_works() {
+    /// This is a test function
+    #[docs_const]
+    #[allow(dead_code)]
+    fn test_fn(input: String) -> String {
+        input.to_ascii_lowercase()
+    }
+
+    assert_eq!(TEST_FN_DOCS, "This is a test function");
+}
+
+#[test]
+fn generics_and_lifetime_works() {
+    /// This is a test function
+    #[docs_const]
+    #[allow(dead_code)]
+    #[allow(clippy::extra_unused_lifetimes)]
+    fn test_fn<'a, T: std::fmt::Debug>(input: T) -> T {
+        input
+    }
+
+    assert_eq!(TEST_FN_DOCS, "This is a test function");
+}
+
```
Honestly, I'm not that concerned about those not working. After all we're just putting the item back verbatim.

cyqsimon commented on documented-test/src/attr/docs_const.rs:
```diff
@@ -29,6 +29,20 @@ fn multiple_docs_work() {
+            #[docs_const(vis = pub)]
+            #[allow(dead_code)]
+            /// Arjun!
+            trait RandomStudent {}
+        }
+
+        assert_eq!(class::RANDOMSTUDENT_DOCS, "Arjun!");
```
I think it may be prudent to use a proper case conversion library like [convert_case](https://crates.io/crates/convert_case), just so that things like `PascalCase` or `camelCase` (which I don't think is used in Rust at the moment, but it doesn't hurt) are correctly handled.

cyqsimon commented on documented-macros/src/lib.rs:
```diff
@@ -407,38 +410,95 @@ pub fn documented_variants(input: TokenStream) -> TokenStream {
+    let docs = match get_docs(attrs, config.trim) {
+        Ok(Some(docs)) => docs,
+        Ok(None) => {
+            // IDEA: customisation: allow_empty
```
This option would be applicable to the derive macros as well, so I'll leave this here for now.

cyqsimon commented:
@SeseMueller I'm pretty happy with the current state of things right now. Can you please take a look to see if you're happy too? Thanks.

SeseMueller commented:
Looks very good! I didn't know the parse_macro_input macro could parse to item, so that's a very nice solution. 
This also eliminates the need to test all the different Item variants because all (that are supported) are handled the same way. 
Great work! 
I can't think of anything that could reasonable be added at this point of time. 
I'm happy and would love to see this merged.

cyqsimon commented:
Alright! A couple more feature gate related shenanigan fixes then I'll merge.
